tests/abi/tools/revng-ensure-rft-equivalence/Main.cpp

```diff
@@ -74,60 +103,159 @@ int main(int Argc, char *Argv[]) {
+    auto *LeftStack = Left->StackArgumentsType().UnqualifiedType().get();
+    auto *RightStack = Right->StackArgumentsType().UnqualifiedType().get();
 
     // XOR the `bool`eans - make sure that either both functions have stack
     // argument or neither one does.
-    revng_check(!LeftStackArguments == !RightStackArguments);
+    revng_check(!LeftStack == !RightStack);
 
     // Ignore function pairs without stack arguments.
-    if (LeftStackArguments == nullptr)
+    if (LeftStack == nullptr)
       continue;
 
-    // If IDs differ - replace the ID.
-    if (LeftStackArguments->ID() != RightStackArguments->ID()) {
-      model::TypePath FromPath = Right->StackArgumentsType().UnqualifiedType();
+    if (auto R = ensureMatchingIDs(LeftStack->key(), RightStack->key(), Binary))
+      RightReplacements.emplace(std::move(R.value()));
+  }
 
-      RightModel->Model->Types().erase(LeftStackArguments->key());
-      auto *Struct = llvm::dyn_cast<model::StructType>(RightStackArguments);
-      revng_check(Struct != nullptr);
-      auto Copy = model::UpcastableType::make<model::StructType>(*Struct);
-      Copy->ID() = LeftStackArguments->ID();
-      auto ToPath = RightModel->Model->recordNewType(std::move(Copy));
+  // Gather all the pure function pointers, as in the types without
+  // a `model::Function` using them.
+  std::vector<const model::RawFunctionType *> LeftFunctionPointerTypes;
+  for (const auto &Left : LeftModel->Model->Types())
+    if (auto *Raw = llvm::dyn_cast<model::RawFunctionType>(Left.get()))
+      if (!FunctionIDLookup.contains(Raw->ID()))
+        LeftFunctionPointerTypes.emplace_back(Raw);
+  std::vector<const model::RawFunctionType *> RightFunctionPointerTypes;
+  for (const auto &Right : RightModel->Model->Types())
+    if (auto *Raw = llvm::dyn_cast<model::RawFunctionType>(Right.get()))
+      if (!FunctionIDLookup.contains(Raw->ID()))
+        RightFunctionPointerTypes.emplace_back(Raw);
 
-      Replacements.emplace(FromPath, ToPath);
-    }
+  // Remove these function pointers (and replace them with `void *`s) to avoid
+  // them cluttering the diff result, since we can only reliably test real
+  // functions (only ones with ABIArtifact available at that).
+  model::TypePath Void;
+  Void = LeftModel->Model->getPrimitiveType(model::PrimitiveTypeKind::Void, 0);
+  for (const model::RawFunctionType *Type : LeftFunctionPointerTypes) {
+    model::TypePath Reference = LeftModel->Model->getTypePath(Type->key());
+    LeftReplacements.emplace(Reference, Void);
+    LeftModel->Model->Types().erase(Type->key());
+  }
+  Void = RightModel->Model->getPrimitiveType(model::PrimitiveTypeKind::Void, 0);
+  for (const model::RawFunctionType *Type : RightFunctionPointerTypes) {
+    model::TypePath Reference = RightModel->Model->getTypePath(Type->key());
+    RightReplacements.emplace(Reference, Void);
+    RightModel->Model->Types().erase(Type->key());
   }
```

Comment:
`purgeUnnamedAndUnreachableTypes` should take care of this.

Reply:
I'll give it a try.

Reply:
No it does not.

They are used as function pointers - so they are referenced, in struct fields and such.

But they interfere with the test (I have no way to tell which is which after the conversion - because we track functions using their `OriginalName`s now) - hence it's easier to remove them over dealing with the repercussions.

Reply:
Function pointers in `struct`s? Where do they come from?

Are they from the libc? Do we need to import the libc or can we start from something less noisy to being with?

Reply:
1. Dwarf introduces them.

2. We need to link against libc for `prinf` and friends :sadparrot:

Reply:
Link != import symbols for dynamic libraries. But I guess the compiler injects this stuff in the main binary's DWARF.

Can we whitelist types by name (e.g., a prefix) and wipe out all the rest?
If not, this is OK.

Reply:
TBH, I'd rather remove this stuff here, where there could be no other consequences (because both models are discarded right after they are `diff`'ed a couple of lines down) rather than try to setup more explicit pruning of the models by making types even more name-dependent.

Reply:
Who references these guys? It's probably some `DynamicFunction`, which you can easily purge.
Then `purgeUnnamedAndUnreachableTypes` should do its job.

Reply:
It's a struct passed as an argument somewhere, if my memory serves me right. But IMO, I'd rather go the "remove all the function pointers" route over "remove that specific named dynamic function", because it's more binary agnostic - no changes to compiler/libc/whatever can break it.

Reply:
"Remove function pointers" is quite more ad-hoc than "remove all dynamic functions".
Also, making sure we purge *everything* it's unrelated should ease debugging.

In this test we're interested exclusively on prototypes of a set of well known, hand written `model::Function`s.
Anything else is noise and should get purged.

Reply:
I tried this out, make the binary dynamically linked and removed all the dynamic functions - the problem is still there.

The subtrees leading to the problem function pointers are not really connected to any functions, it seems. BUT they are not getting "purged" because they are named.

(also when linked dynamically, the offenders are different structs 🤔, which makes me even more concerned about any "soft" approaches: when linked statically, `pthread`-related stuff was the problem, when dynamically - it's `FILE` and its internals)

Reply:
Ah, wait, I just realized there's a (relatively) new `pruneUnusedTypes`. Which seems to do exactly what was needed.

Reply:
Please take a look at the final state of `revng-ensure-rft-equivalence/Main.cpp` and feel free to close this thread if it looks good.